### PR TITLE
Batch processing. Validation of HXN support.

### PR DESCRIPTION
**Pull request by dmgav**
This PR involves the following changes to the code:

-- Refactoring of ``pyxrf_batch`` function (may be called from IPython or from custom scripts).

-- Testing PyXRF on data from HXN beamline.

-- Validation of batch processing routines (``make_hdf`` and ``pyxrf_batch`` on data from HXN and SRX beamlines.

### CODE REVIEW

File: pyxrf/model/command_tools.py [hunk 1]
```diff
@@ -208,40 +219,179 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+                print(f"File with Scan ID {start_id} was not found.")
+            else:
+                print(f"Processing file with Scan ID {start_id}")
+        else:
+            # ``start_id`` and ``end_id`` are specified:
+            #   select files, which contain the respective ID substring in their names
+            flist = []
+            for data_id in range(start_id, end_id+1):
+                flist += [fname for fname in all_files
+                          if re.search(f"^[^_]*_{str(data_id)}\D+", fname)]  # noqa: W605
```
Comment (mrakitin) [inline pyxrf/model/command_tools.py]: This regex-string can be specified somewhere above only once, and reused here.

File: pyxrf/model/command_tools.py [hunk 2]
```diff
@@ -208,40 +219,179 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+            if len(flist) < 1:
+                print(f"No files with Scan IDs in the range {start_id} .. {end_id} were not found.")
```
Comment (mrakitin) [inline pyxrf/model/command_tools.py]:
```suggestion
                print(f"No files with Scan IDs in the range {start_id} .. {end_id} were found.")
```

File: pyxrf/model/command_tools.py [hunk 3]
```diff
@@ -208,40 +219,179 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+            else:
+                print(f"Processing file with Scan IDs in the range {start_id} .. {end_id}")
+
+    # Check if .json parameter file exists
+    if not os.path.isabs(param_file_name):
+        pname = os.path.join(wd, param_file_name)
+    else:
+        pname = param_file_name
+    if not os.path.exists(pname) and not os.path.isfile(pname):
```
Comment (mrakitin) [inline pyxrf/model/command_tools.py]:
```suggestion
    if not os.path.exists(pname) or not os.path.isfile(pname):
```

File: pyxrf/model/command_tools.py [hunk 4]
```diff
@@ -208,40 +219,179 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+        print(f"ERROR: can't find the parameter file {pname}. "
+              "Check the value of 'param_file_name'.")
+        return 1
```
Comment (mrakitin) [inline pyxrf/model/command_tools.py]: Again, I'd recommend to use `raise` here.

File: pyxrf/model/command_tools.py [hunk 5]
```diff
@@ -208,40 +219,179 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+        print(f"Parameter file: {pname}")
+
+    if len(flist) > 0:
+
+        print(f"The following files are scheduled for processing:")
+        for fln in flist:
+            print(f"    {fln}")
+        print(f"Total number of selected files: {len(flist)}")
+        print()
+        print("Processing ...")
```
Comment (mrakitin) [inline pyxrf/model/command_tools.py]: We can eliminate the extra prints by using `print("\nProcessing...\n")`.
```suggestion
        print("\nProcessing ...\n")
```

File: pyxrf/model/command_tools.py (hunk 1 above)
Comment (dmgav) [reply pyxrf/model/command_tools.py]: Regex string depends on ``data_id``, which is changing in the loop.

File: pyxrf/model/command_tools.py [hunk 6]
```diff
@@ -208,40 +219,179 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+            param_file_name="~/data_parameters/parameters.json"
+
+    How does ``pyxrf-batch`` identify a file with a given Scan ID?
+    In order for the batch mode processing to work correctly, the data file name must
+    conform to the following structure requirements:
+
+        <prefix>_<scanID>(any non-digit character)(any character sequence).h5
+
+    <prefix> - any sequence of characters, not containing ``_``
+    <scanID> - string version of Scan ID, obtained as ``str(id)``
```
[the remark this answers is not in the document]
Comment (dmgav) [reply pyxrf/model/command_tools.py]: Fixed

File: pyxrf/model/command_tools.py [hunk 7]
```diff
@@ -208,40 +219,179 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+            data_files_valid = True
+            for fln in data_files:
+                if not isinstance(fln, str):
+                    data_files_valid = False
+                    break
+
+        if not data_files_valid:
+            print(f"ERROR: the list of data files has invalid format. Check the parameter 'data_files'.")
+            print(f"   data_file = {data_files}")
+            return 1
```
[the remark this answers is not in the document]
Comment (dmgav) [reply pyxrf/model/command_tools.py]: Fixed

File: pyxrf/model/command_tools.py (hunk 3 above)
Comment (dmgav) [reply pyxrf/model/command_tools.py]: Fixed

File: pyxrf/model/command_tools.py (hunk 4 above)
Comment (dmgav) [reply pyxrf/model/command_tools.py]: Fixed

File: pyxrf/model/command_tools.py [hunk 8]
```diff
@@ -208,40 +223,192 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+            data_files_valid = True
+            for fln in data_files:
+                if not isinstance(fln, str):
+                    data_files_valid = False
+                    break
+
+        if not data_files_valid:
+            print(f"ERROR: the list of data files has invalid format. Check the parameter 'data_files'.")
+            print(f"   data_files = {data_files}")
+            raise ValueError("Function pyxrf_batch. Parameter 'data_files' has invalid format.")
```
Comment (mrakitin) [inline pyxrf/model/command_tools.py]: Thanks. I don't think we need prints anymore as the string can be added to the `ValueError`.

File: pyxrf/model/command_tools.py [hunk 9]
```diff
@@ -208,40 +219,179 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+
+        # At this point ``data_files`` is a list of str.
+        for n, fln in enumerate(data_files):
+            data_files[n] = os.path.expanduser(fln)
+
+        # Working directory name is appended to each file name in the list,
+        #     which does not contain full path
+        for n, fln in enumerate(data_files):
+            if not os.path.isabs(fln):  # Check if ``fln`` is an absolute file path
+                data_files[n] = os.path.join(wd, fln)
```
[the remark this answers is not in the document]
Comment (mrakitin) [reply pyxrf/model/command_tools.py]: Please also address https://github.com/NSLS-II/PyXRF/pull/190/files#r330294353.

File: pyxrf/model/command_tools.py (hunk 1 above)
Comment (mrakitin) [reply pyxrf/model/command_tools.py]: OK, makes sense to leave it as is, to avoid extra complexity.

File: pyxrf/model/command_tools.py [hunk 10]
```diff
@@ -208,40 +223,192 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+
+    # Check if .json parameter file exists
+    if not os.path.isabs(param_file_name):
+        pname = os.path.join(wd, param_file_name)
+    else:
+        pname = param_file_name
+    if not os.path.exists(pname) or not os.path.isfile(pname):
+        print(f"ERROR: can't find the parameter file {pname}. "
+              "Check the value of 'param_file_name'.")
+        raise IOError(f"Function 'pyxrf_batch'. Processing parameter file '{pname}' does not exist.")
```
Comment (mrakitin) [inline pyxrf/model/command_tools.py]: The `print(...)` above can be eliminated.

File: pyxrf/model/command_tools.py [hunk 11]
```diff
@@ -208,40 +223,192 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
             fit_pixel_data_and_save(working_directory, fname,
                                     fit_channel_sum=fit_channel_sum, param_file_name=param_file_name,
                                     fit_channel_each=fit_channel_each, param_channel_list=param_channel_list,
                                     incident_energy=incident_energy, spectrum_cut=spectrum_cut,
                                     save_txt=save_txt, save_tiff=save_tiff,
-                                    ic_name=ic_name)
+                                    ic_name=ic_name, use_average=use_average)
+
+        print()
+        print("All selected files were processed.")
```
Comment (mrakitin) [inline pyxrf/model/command_tools.py]: Maybe better to use one print with `\n`?

File: pyxrf/model/command_tools.py [hunk 12]
```diff
@@ -208,40 +219,179 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+    <scanID> - string version of Scan ID, obtained as ``str(ScanID)``
+    <prefix> and <scanID> are separated by ``_``
+    <scanID> is separated from the rest of the file name from the right by any non-number character.
+    Files must have extension .h5. (Note, that ``.`` separating the extension is a non-number character
+    that may terminate the Scan ID part of the file name
+
+    For example, the following files will be recognized as Scan ID 28355:
+
+        scan2D_28355.h5
+        scanNewName_28355_some_comments.h5
```
[the remark this answers is not in the document]
Comment (dmgav) [reply pyxrf/model/command_tools.py]: Actually I would prefer not to include undesired file names in examples even if the program can handle them. Examples are visual references.

File: pyxrf/model/command_tools.py (hunk 8 above)
Comment (dmgav) [reply pyxrf/model/command_tools.py]: I am not sure. Typical way to deal with exceptions in home-made script is to catch them and do nothing. It is easier to for the user to find the cause of error.

File: pyxrf/model/command_tools.py (hunk 11 above)
Comment (dmgav) [reply pyxrf/model/command_tools.py]: I am not sure. I always used to think, that empty prints are easier to find, insert, delete and comment out in the code, especially if there is a long sequence of prints. Empty prints show the locations of empty lines in the output.

File: pyxrf/model/command_tools.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mrakitin) [reply pyxrf/model/command_tools.py]: OK, sounds reasonable.

File: pyxrf/model/command_tools.py [hunk 13]
```diff
@@ -208,40 +223,187 @@ def pyxrf_batch(start_id, end_id=None, wd=None, fit_channel_sum=True, param_file
...
+            data_files = list(data_files)  # Convert to list
+            data_files_valid = True
+            for fln in data_files:
+                if not isinstance(fln, str):
+                    data_files_valid = False
+                    break
+
+        if not data_files_valid:
+            raise ValueError("Function pyxrf_batch. Parameter 'data_files' has invalid format. \n"
+                             "data_files = '{data_files}'")
```
Comment (mrakitin) [inline pyxrf/model/command_tools.py]: f-string is missing here.

File: pyxrf/model/command_tools.py (hunk 10 above)
Comment (mrakitin) [reply pyxrf/model/command_tools.py]: Thanks.

**Pull request merged**